query-engine/core/src/query_ast/read.rs

```diff
@@ -86,6 +86,39 @@ impl Display for ReadQuery {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum QueryOption {
+    ThrowOnEmpty,
+}
+pub const THROW_ON_EMPTY: [QueryOption; 1] = [QueryOption::ThrowOnEmpty];
```

Comment:
This is to allow queries to have options, and express things like whether when it executes, it throws on empty result set or not. A simple boolean flag could have made it, and maybe YAGNI more options, but I thought the cost wasn't too high, and using a boolean flag would have been too concrete. 

It also influenced my decision to wrap a vector (despite the allocation) the fact that both `RecordQuery` and `ManyRecordsQuery` owned their fields, otherwise, I would have made QueryOptions be `pub struct QueryOptions<'a>(pub &'a [QueryOption]);`

Reply:
Please let me know if there are good reasons to make this just a flag, or opt for a different design.

Reply:
I tend not to overthink the future too much in those cases and make such refactors only when we realize we actually need to. Might be that in 4 years, we won't need anything else than this simple boolean and this will feel overkill. Might be that in 2 months we'll need more, at which point doing the refactor won't be very costly anyway but will make a lot of sense. It _doesn't_ mean this is bad though, this is just a personal way of thinking. Feel free to keep it that way!

Another suggestion could've been to use a [bitflag instead of an array](https://github.com/prisma/prisma-engines/blob/156d8b51290f79bcfc4125c80a9e63c4e03fb102/psl/psl-core/src/common/preview_features.rs#L77-L121), which makes it faster to check whether an option is used. But then again, unless this check is made 100 times in a query lifecycle and it becomes a bottleneck, this is just premature optimization.

To be clear, it really doesn't make any difference here, this is just to share knowledge on what you can do in Rust, if you aren't aware already. So don't feel like you need to change anything, this is all fine 👍

Reply:
I tend not to overthink the future too much in those cases and make such refactors only when we realize we actually need to. Might be that in 4 years, we won't need anything else than a simple boolean and this will feel overkill. Might be that in 2 months we'll need more, at which point doing the refactor won't be very costly anyway but will make a lot of sense. It _doesn't_ mean this is bad though, this is just a personal way of thinking. Feel free to keep it that way!

Another suggestion could've been to use a [bitflag instead of an array](https://github.com/prisma/prisma-engines/blob/156d8b51290f79bcfc4125c80a9e63c4e03fb102/psl/psl-core/src/common/preview_features.rs#L77-L121), which makes it faster to check whether an option is used. But then again, unless this check is made 100 times in a query lifecycle and it becomes a bottleneck, this is just premature optimization.

To be clear, it really doesn't make any difference here, this is just to share knowledge on what you can do in Rust, if you aren't aware already. So don't feel like you need to change anything, this is all fine 👍

Reply:
Aggreed, I followed symmetry on the existing structs for the queries, but I don't find it necessary either. Another reason I modeled it this way too, is that an empty vector is zero-cost abstraction, which will be the case for most queries. But you are right, it could very well work with a flag, period. (And I would probably have done it that way if started over, no strong opinion)

Thanks for the reference to the bitflag, it is a much better alternative to an array!

Reply:
Feedback was addressed in https://github.com/prisma/prisma-engines/pull/3356/commits/c2330f61320a2289d3e36b0c3d30d66e7eeb8bb2